docs/map.rst

```diff
@@ -0,0 +1,629 @@
+
+
+Importing Map and IntMap
+------------------------
+
+When using ``Map`` or ``IntMap`` in a Haskell source file you should always use
+a ``qualified`` import because these modules export names that clash with the
+standard Prelude (you can import the type constructor on its own though!).
+
+::
+
+    import Data.Map.Strict (Map)
+    import qualified Data.Map.Strict as Map
+
+    import Data.IntMap (IntMap)
+    import qualified Data.IntMap as IntMap
+
+
+Common API Functions
+--------------------
+
+.. TIP::
+   All of these functions that work for ``Map`` will also work for ``IntMap``,
+   which has the key type ``k`` specialized to ``Int``. Anywhere that you
+   see ``Map k v`` you can replace it with ``IntMap v``.
+
+.. NOTE::
+   ``fromList [some,map,entries]`` is how a ``Map`` is printed.
+
+
+Construction and Conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Create an empty map
+"""""""""""""""""""
+
+::
+
+    Map.empty :: Map k v
+    Map.empty = ...
+
+:map:`empty` creates a map with zero entries.
+
+::
+
+    Map.empty
+    > fromList []
+
+Create a map with one entry (singleton)
+"""""""""""""""""""""""""""""""""""""""
+
+::
+
+    Map.singleton :: k -> v -> Map k v
+    Map.singleton key value = ...
+
+:map:`singleton` creates a map with a single ``(key,value)`` entry in it.
+
+::
+
+    Map.singleton 1 "one"
+    > fromList [(1,"one")]
+
+    Map.singleton "containers" ["base"]
+    > fromList [("containers",["base"])]
+
+Create a map from a list
+""""""""""""""""""""""""
+
+::
+
+    Map.fromList :: Ord k => [(k, v)] -> Map k v
+    Map.fromList xs = ...
+
+:map:`fromList` creates a map containing the entries of the list ``xs`` where
+the keys comes from the first entries of the pairs and the values from the
+second. If the same key appears more than once then the last value is taken.
+
+::
+
+    Map.fromList []
+    > fromList []
+
+    Map.fromList [(1,"uno"), (1,"one"), (2,"two"), (3,"three")]
+    > fromList [(1,"one"),(2,"two"),(3,"three")]
+
+There's another incredibly useful function for constructing a map from a list::
+
+    Map.fromListWith :: Ord k => (a -> a -> a) -> [(k, a)] -> Map.Map k a
+    Map.fromListWith f xs = ...
+
+:map:`fromListWith` allows you to build a map from a list ``xs`` with repeated
+keys, where ``f`` is used to "combine" (or "choose") values with the same key.
+
+::
+
+    -- Build a map from a list, but only keep the largest value for each key.
+    Map.fromListWith max [("a", 1), ("a", 2), ("b", 2)]
+    > fromList [("a",2),("b",2)]
+
+    -- Build a histogram from a list of elements.
+    Map.fromListWith (+) (map (\x -> (x, 1)) ["a", "a", "b", "c", "c", "c"])
+    > fromList [("a",2),("b",1),("c",3)]
+
+
+Create a list from a map
+""""""""""""""""""""""""
+
+::
+
+    Map.elems :: Map k v -> [v]
+    Map.elems m = ...
+
+:map:`elems` returns a list of values held in the map ``m``.
+
+::
+
+    Map.toAscList, Map.toList, Map.assocs :: Map k v -> [(k, v)]
+    Map.toAscList m = ...
```

Comment:
Done.

Reply:
I'd mention that `Foldable.toList` is rarely useful for maps, but that its functionality is also available from `elems`. Or something like that. The goal is to avoid confusion, not to steer people toward things that aren't too relevant.

Reply:
Thinking about this more I'm leaning towards just mentioning toAscList and excluding the others (as well as the note about toList). WDYT?

Reply:
I think the note about `toList` should be included, because it's a potential gotcha.

Reply:
Done.